### ISSUE TITLE
sqlite3 option for org-roam-database-connector does not delete or update the database cache

### ISSUE THREAD

**Issue by nobiot** [org-roam/org-roam#1927]
### Description
Set `(setq org-roam-database-connector 'sqlite3)`. Org-roam caches a new node created.
It does not update or delete an existing node -- so C and R of CRUD work but U and D don't.

#### Steps to Reproduce

0. Delete configuration files (remove the directory `~/.config/emacs`)
1. Load Emacs; ensure it is loaded with no configuration
2. Set `package-archives` with MELPA and install `org-roam`, its dependecies, and `emacsql-sqlite3`
3. Ensure `sqlite3` is installed
4. Create a new file node in `org-roam-directory` -- e.g. first.org
5. Create a second file node -- e.g. second.org
6. Confirm both nodes appear with `org-roam-node-find`
7. Update second file node -- e.g. `org-roam-node-insert` to link to first.org
8. Save
9. See the screen shot

![Screenshot from 2021-10-27 17-41-57](https://user-images.githubusercontent.com/12507865/139133419-c3d18dfe-7ece-4346-ad1a-a963243b5fcd.png)

10. Use Dired and delete the second file
11. Run `org-roam-db-sync`
12. Run `org-roam-node-find`; you will see the second file still in the list of completion candidates

![Screenshot from 2021-10-27 17-43-25](https://user-images.githubusercontent.com/12507865/139134639-688f7b52-b983-423c-9b53-97bce384e0d3.png)

#### Backtrace
For update error:
```
Debugger entered--Lisp error: (emacsql-error "near line 331: UNIQUE constraint failed: nodes.id")
  signal(emacsql-error ("near line 331: UNIQUE c<!-- Example: File A is there -->onstraint failed: nodes.id"))
  #f(compiled-function (conn) #<bytecode -0x1a47a8250eaf33be>)(#<emacsql-sqlite3-connection emacsql-sqlite3-connection-155ef80413ce>)
  apply(#f(compiled-function (conn) #<bytecode -0x1a47a8250eaf33be>) #<emacsql-sqlite3-connection emacsql-sqlite3-connection-155ef80413ce> nil)
  emacsql-parse(#<emacsql-sqlite3-connection emacsql-sqlite3-connection-155ef80413ce>)
  #f(compiled-function (connection sql &rest args) "Send SQL s-expression to CONNECTION and return the results." #<bytecode -0x1df356467015a8fb>)(#<emacsql-sqlite3-connection emacsql-sqlite3-connection-155ef80413ce> [:insert :into nodes :values $v1] (["77f8789d-9b4b-456d-af4a-ba80198b7671" "/home/nobiot/org-roam/20211027212918-second.org" 0 1 nil nil nil nil "second" (("CATEGORY" . "20211027212918-second") ("ID" . "77f8789d-9b4b-456d-af4a-ba80198b7671") ("BLOCKED" . "") ("FILE" . "/home/nobiot/org-roam/20211027212918-second.org") ("PRIORITY" . "B")) nil]))
  apply(#f(compiled-function (connection sql &rest args) "Send SQL s-expression to CONNECTION and return the results." #<bytecode -0x1df356467015a8fb>) #<emacsql-sqlite3-connection emacsql-sqlite3-connection-155ef80413ce> ([:insert :into nodes :values $v1] (["77f8789d-9b4b-456d-af4a-ba80198b7671" "/home/nobiot/org-roam/20211027212918-second.org" 0 1 nil nil nil nil "second" (("CATEGORY" . "20211027212918-second") ("ID" . "77f8789d-9b4b-456d-af4a-ba80198b7671") ("BLOCKED" . "") ("FILE" . "/home/nobiot/org-roam/20211027212918-second.org") ("PRIORITY" . "B")) nil])))
  emacsql(#<emacsql-sqlite3-connection emacsql-sqlite3-connection-155ef80413ce> [:insert :into nodes :values $v1] (["77f8789d-9b4b-456d-af4a-ba80198b7671" "/home/nobiot/org-roam/20211027212918-second.org" 0 1 nil nil nil nil "second" (("CATEGORY" . "20211027212918-second") ("ID" . "77f8789d-9b4b-456d-af4a-ba80198b7671") ("BLOCKED" . "") ("FILE" . "/home/nobiot/org-roam/20211027212918-second.org") ("PRIORITY" . "B")) nil]))
  org-roam-db-query([:insert :into nodes :values $v1] (["77f8789d-9b4b-456d-af4a-ba80198b7671" "/home/nobiot/org-roam/20211027212918-second.org" 0 1 nil nil nil nil "second" (("CATEGORY" . "20211027212918-second") ("ID" . "77f8789d-9b4b-456d-af4a-ba80198b7671") ("BLOCKED" . "") ("FILE" . "/home/nobiot/org-roam/20211027212918-second.org") ("PRIORITY" . "B")) nil]))
  org-roam-db-query!(#f(compiled-function (err) #<bytecode -0x1004ab219f8b4cb7>) [:insert :into nodes :values $v1] ["77f8789d-9b4b-456d-af4a-ba80198b7671" "/home/nobiot/org-roam/20211027212918-second.org" 0 1 nil nil nil nil "second" (("CATEGORY" . "20211027212918-second") ("ID" . "77f8789d-9b4b-456d-af4a-ba80198b7671") ("BLOCKED" . "") ("FILE" . "/home/nobiot/org-roam/20211027212918-second.org") ("PRIORITY" . "B")) nil])
  org-roam-db-insert-file-node()
  org-roam-db-update-file()
  org-roam-db-autosync--try-update-on-save-h()
  run-hooks(after-save-hook)
  basic-save-buffer(t)
  save-buffer(1)
  funcall-interactively(save-buffer 1)
  command-execute(save-buffer)
```

#### Expected Results

No error when updating the second node.
When the node is deleted, the cache gets updated correctly.

#### Actual Results

As above. You get an error on update.
Delete does not update the cache.

### Environment

- Emacs: GNU Emacs 28.0.50 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.25, cairo version 1.16.0)
 of 2021-09-20
- Framework: 
- Org: Org mode version 9.4.4 (release_9.4.4 @ /usr/local/share/emacs/28.0.50/lisp/org/)
- Org-roam: 2.1.0
- org-roam-20211027.1328 (MELPA)

I also observed the same issue with Emacs on Windows (Emacs 27.2, Org 9.5).

**Comment by jethrokuan** [comment]
This is a duplicate of #1910 . I've looked into this and don't have a solution for this, but I can at lay out the issue so maybe someone can find a proper fix.

Org-roam uses `foreign keys` to cascade updates across the table. We want deletes from the `files` table to propagate to the `nodes` table, and `node` deletions to propagate to the `links`, `citations` table etc. The `sqlite` option enables foreign keys by default. For all other connector options, we try to do it according to the [sqlite3 manual](https://sqlite.org/foreignkeys.html):

https://github.com/org-roam/org-roam/blob/1b221a1d4ad0edae746a7eae435f4c3bca56c10e/org-roam-db.el#L242

Upon manual inspection of the database, I find that foreign keys support is not correctly enabled (still 0).  The sqlite3 manual says that this cannot be enabled in a transaction, which is what code in master is currently doing and should therefore be changed. But I've tried moving the sql query outside a transaction, and it doesn't seem to have any effect.

```emacs-lisp
(emacsql (org-roam-db) "PRAGMA foreign_keys = ON")
```

You can try running the above snippet of code anywhere and it doesn't seem to have any effect for me.

Does not correctly turn on foreign key support. However, if I manually run `sqlite3 ~/.emacs.d/org-roam.db` and run `PRAGMA foreign_keys = ON;` then things work fine and the errors will disappear. Would appreciate if someone could find find a solution here.

**Comment by jethrokuan** [comment]
Should be fixed with #2009

**Comment by nobiot** [comment]
@jethrokuan 
Confirm that #2009 works on Windows 10.
Both create and update work.

The original post was done with Ubuntu but I'm fine as long as Windows works for this. I might get to testing Ubuntu later.
For now, I'm happy to close the issue.


# An important finding

Ensure to delete the `org-roam.db` file created with other DB connector; otherwise, Org-roam will likel output an "UNIQUE constraint failed" error. I originally had this error; deleting the db file and building it again with `sqlite3` fixed it.

# Environment and procedure of my tests in detail below

  - Tested with
    + Windows 10
    + Emacs: GNU Emacs 27.2 (build 1, x86_64-w64-mingw32)
 of 2021-03-26
    + Org: Org mode version 9.4.4 (release_9.4.4 @ c:/Program Files/Emacs/share/emacs/27.2/lisp/org/)
    + emacsql-sqlite3-20200914.508
    + org-roam-20211221.623

  
  - Download and expand sqlite-tools-win32-x86-3370000.zip (1.84 MiB)
    It's the latest release, Version 3.37.0 (2021-11-27)
    https://sqlite.org/download.html

  - Place `sqlite3.exe` in `C:\Users\nobiot\AppData\Roaming\bin`

  - Add PATH `C:\Users\nobiot\AppData\Roaming\bin` in Environmental Variable

  - Confirm `sqlite --version` works in `cmd`

  - Launch Emacs (after the PATH added so that Emacs recognises it)

  - Ensure config contains the following:

``` emacs-lisp
;; Connector needs to be set before autosync-mode and db-sync
;; Otherwise, the binary not found error
(setq org-roam-database-connector 'sqlite3)
(org-roam-db-autosync-mode)
```

  - Test Org-roam as originally posted (create two new files, and insert a link from one to another).

**Comment by nobiot** [comment]
Just as a way to keep track of this issue.
Windows is fine but Ubuntu isn't (!)
See https://github.com/org-roam/org-roam/pull/2009#issuecomment-998720049

**Comment by jueqingsizhe66** [comment]
Windows 10
Emacs: GNU Emacs 27.2 
of 2021-03-26
Org: Org mode version 9.4.6 
emacsql-sqlite3-20220304.1014
org-roam-20220515.2351

2022-05-16, I tried to insert my org-roam notes, and found that org-roam not working.   

> I checked the org-roam.db, and without update ;( tried to org-roam-db-syn, all the org-roam.db has been update with empty data( org-roam cannot get filles in the org-roam directories)
> I tried to find out the date of last working file in my  org-roam directory:   2022-04-23,
> I have tried ideas above , thanks, but not working. Maybe  I should use the org-roam older version instead.
> Love org-roam!

**Comment by jueqingsizhe66** [comment]
I just use  org-roam-20220121.2350 instead of org-roam-20220515.2351 , and it works without the care of emacs-sqlite.exe or sqlites etc.

Now  i can use org-roam-find-file and org-roam-insert-file again , and i found that the org-roam.db is also updated with my operation, works again.

**Comment by nobiot** [comment]
@jethrokuan  it looks like your #2185 overrode the fix #2178 (`relative-path`).
@jueqingsizhe66 I think your issue is caused by the fact your org-roam files contain `data` somewhere in the path.
For now you could customize `org-roam-file-exclude-regexp`.
